lp1716844: Escaping control characters in the audit log

dutow opened the pull request:
Implemented escaping for the JSON output format, and updated the test suite.

The other output formats are left unchanged, as they do not have a well defined method for escaping these characters.
XML does support control characters in version 1.1, but most tools only understand 1.0, and our output is 1.0.

percona-ysorokin commented on plugin/audit_log/audit_log.c:
```diff
@@ -238,6 +271,39 @@ void json_escape(const char *in, size_t *inlen, char *out, size_t *outlen)
     { '\f',  2, "\\f" },
     { '\t',  2, "\\t" },
+    // escaping basic ASCII control characters
+    // skipping 0: it can't be handled by the current algorithm
```
The same comment, plus
`\b` and `\f`

percona-ysorokin commented on plugin/audit_log/audit_log.c:
```diff
@@ -219,6 +219,39 @@ void xml_escape(const char *in, size_t *inlen, char *out, size_t *outlen)
+    { 8,  8, "\\u#x0006" },
+    { 7,  8, "\\u#x0007" },
+    { 8,  8, "\\u#x0008" },
+    { 9,  8, "\\u#x0009" },
```
Is this OK that we have duplicates in this table?
`'\t'` == `9`
`'\n'` == `10`
`'\r'` == `13`

dutow replied:
I didn't want to delete the original entries (to reduce the chance of merge conflicts later), but I also thought that adding every entry to this list would be clearer than skipping those records.

Technically, it won't cause any problems, as the escape code will stop as soon as it finds an entry in this list. If I remove one of the variants, it could end up a bit faster, but now we are dealing with quite larger arrays than before (around 40 entries vs the 7-8 original). 

If this is a performance critical part of the server (I thought it isn't that critical, as it looks like a verbose logging feature), I could implement this differently, but that wouldn't be just the adding of a few lines to the existing escaping code, but changing parts of the actual escape loop.

laurynas-biveinis replied:
This is fully our code. Merge conflicts should not be a concern

dutow replied:
In this case, I updated the PR based on my original idea: I introduced a new array for the control characters, so we won't end up in a 40 element loop for each character.

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -219,6 +219,39 @@ void xml_escape(const char *in, size_t *inlen, char *out, size_t *outlen)
+    { 3,  8, "\\u#x0003" },
+    { 4,  8, "\\u#x0004" },
+    { 5,  8, "\\u#x0005" },
+    { 8,  8, "\\u#x0006" },
```
6 ?

dutow replied:
Fixed.

laurynas-biveinis commented on mysql-test/include/audit_log_events.inc:
```diff
@@ -6,6 +6,8 @@ CREATE TABLE t1
               c2 CHAR(20));
 INSERT INTO t1 VALUES (1,'a'),(2,'b'),(3,'c');
+INSERT INTO `t1` VALUES (4,NULL);
+INSERT INTO `t1` VALUES (5,'541S?');
```
Probably wouldn't hurt to test the escaped character set exhaustively.

dutow replied:
The tests currently are far from exhaustive (for example, the actual xml or json output isn't tested, only to the point that it is valid), these statements here only verify the SQL statements that were executed when creating the database.

So unless I also add some actual output verification, exhaustive testing wouldn't accomplish much more. Should I do that?

laurynas-biveinis replied:
It's just 31 characters - if sorted in input, recording the manual outputs of \01 \02 ... etc would be enough

dutow commented:
Updates:

* I removed the xml escapes, as that's only supported by XML 1.1, and we are generating 1.0 output. Many tools only support 1.0, so it's probably a good idea to keep using that.
* I added a more extensive test for the json output
* Control character tests are only included in the json output                                     
* Added unsigned casts to a few places, as signed chars caused bugs

dutow commented:
Jenkins run: https://jenkins.percona.com/job/percona-server-5.6-param/2034/

laurynas-biveinis commented on mysql-test/t/audit_log_json.test:
```diff
@@ -7,17 +7,41 @@ SET GLOBAL audit_log_flush=ON;
 --remove_file $MYSQLD_DATADIR/test_audit.log
 SET GLOBAL audit_log_flush=ON;
 
+--let $TEST_CONTROL_CHARS=1;
```
Uppercase vars in MTR become environment vars too, lowercase ones don't. Since you don't need an env var here (e.g. for interacting with inline Perl), use lowercase. This comment on its own would be too minor to force a re-spin of this PR

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -186,16 +187,29 @@ void escape_buf(const char *in, size_t *inlen, char *out, size_t *outlen,
-    {
-      if (*in == rule->character)
-      {
+    if ((unsigned char)(*in) < 32) {
```
Why a typecast is needed here?

dutow replied:
I could eliminate this cast by specifying less than space, but I'm not sure if that would be clearer.

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -186,16 +187,29 @@ void escape_buf(const char *in, size_t *inlen, char *out, size_t *outlen,
-      if (*in == rule->character)
-      {
+    if ((unsigned char)(*in) < 32) {
+      if (control_escape_rules[(unsigned int)*in].character) {
```
Likewise. If this one is because of sign mismatch, is it possible to make the ```in``` arg ```const unsigned char *``` instead?

dutow replied:
Because by default char is signed in gcc. Even if I specify the parameter as unsigned, I have to add this cast somewhere.

laurynas-biveinis commented on plugin/audit_log/audit_log.c:
```diff
@@ -186,16 +187,29 @@ void escape_buf(const char *in, size_t *inlen, char *out, size_t *outlen,
+      {
+        if (*in == rule->character)
+        {
+          if ((outend - out) < (int) rule->length)
```
Why a typecast here? If it has to stay, wouldn't it be more exact to use ptrdiff_t?

laurynas-biveinis commented on mysql-test/t/audit_log_json.test:
```diff
@@ -7,17 +7,41 @@ SET GLOBAL audit_log_flush=ON;
+              $control_char_count = $control_char_count + 1;
+              my $expected = $last_control_char + 1;
+              if ($expected != $minimum_character) {
+                   print "Incorrent control character in output: Expected $expected, got $minimum_character\n";
```
"Incorrect"

dutow commented:
Updated the PR: 
* fixed the typo
* removed some duplicated code I added (more refactoring)
* changed one of the casts (int->ptrdiff_t)

laurynas-biveinis commented:
Have you tested on Jenkins again?

dutow commented:
Yes: https://jenkins.percona.com/job/percona-server-5.6-param/2037/

(I had to change a variable declaration, because we don't compile with C99 on older platforms)

laurynas-biveinis commented:
@dutow, we usually merge when when all version PRs are ready, in this case the 5.7 one.

dutow commented:
Ops, I completely forgot about that one :( I'll open the 5.7 PR too, it's basically the same change.
